Draft Refresh (#76)
tls-attestation/draft-tls-attestation: draft-fossati-tls-attestation.md

The protocol's security relies on the verifiable binding between these two logically separate units of evidence.

## Authentication vs. Attestation

As noted, the protocol supports either combined platform attestation with X.509 certificate authentication, or attestation only.

> Attestation when used alone is vulnerable to identity spoofing attacks, in particular when zero-day attacks exist for a class of hardware. (TODO: reference). Therefore it needs to be combined with traditional authentication, which in the case of TLS takes the form of X.509 certificates.

Author: I assume that provisioning of the auth credentials relies on attestation, and I talk about that later on (security considerations related to the attestation-only mode). I'd rather avoid it here - this is an overview.

We RECOMMEND that regular applications only use the combined mode, which provides the full security guarantees of an authenticated TLS handshake (for the peer/peers being authenticated) as
well as guarantees on platform integrity.

> The attestation-only mode is included in this document for specialized use cases, including initial provisioning of the TLS stack. In these cases, additional security controls SHOULD be provided,

Author: I'm not sure, because we don't specify any of them here, and so a MUST would be meaningless IMO.

> The attestation-only mode is included in this document for specialized use cases, including initial provisioning of the TLS stack. In these cases, additional security controls SHOULD be provided,

Reviewer: Maybe a problem here is that MUST and SHOULD apply to different parts of the text. For that provisioning use-case MUST is better, but that's only given as an example... 

Quoting the definition of SHOULD below. It seems the appropriate word here.
> 3. SHOULD   This word, or the adjective "RECOMMENDED", mean that there
>   may exist valid reasons in particular circumstances to ignore a
>   particular item, but the full implications must be understood and
>   carefully weighed before choosing a different course.

such as hardware-enforced time limitations, or use of platform-level APIs in the case of cloud infrastructure.

## TLS Identity Key (TIK) {#tik}

> A central property of this protocol is that it binds a peer's identity key, the TIK, to the TLS handshake.

Author: I added text that when the TIK is a TEE-resident key, it SHOULD be persistent (i.e., long lived).

> A central property of this protocol is that it binds a peer's identity key, the TIK, to the TLS handshake.

Reviewer: I think depending on the abstraction level from which you look at this, using either TIK or {LTK, EK} can make more sense. e.g., for a TLS library implementer the difference is irrelevant, whereas for a security architect looking to put a system into production it could make the difference between broken and secure.

I'd rather we stick with TIK throughout the document, provided that we make the distinction clear in the terminology section, and we go to town on separating the two cases in the Security considerations part. But we can do that as a follow-up.
Note: I've not checked whether using TIK makes sense everywhere, or whether in some places we need to call out one of the cases.

This applies to the client's identity (TIK-C), the server's identity (TIK-S) or both.

In the standard mode that combines attestation with certificate authentication, the TIK MUST be
the X.509 certificate's end entity key.

> In the attestation-only mode, the TIK is generated by the TEE and is only used for TLS handshakes.

Reviewer 2: The L (long) in LTK is not an absolute concept; it’s relative to the lifetime of the endpoint.  If the TEE lives an ephemeral existence, the TIK is still “long-term” relative to its life span.

It may be used by (specialized) applications to provide continuity of identity between TLS sessions with the same peer. 

# Attestation Extensions

As typical with new features in TLS, the client indicates support for the new
extension in the ClientHello message. The newly introduced extensions allow
remote attestation credentials and nonces to be exchanged. The nonces are used
for guaranteeing freshness of the exchanged evidence when the background check
model is in use. Nonces are not used in the passport model, because the expectation
of freshness is more relaxed and is only governed by the lifetime of the signed
attestation results.

When either the Evidence or the Attestation Results extension is successfully
negotiated, the content of the corresponding Certificate message contains a
payload that is encoded based on the wrapper defined in {{-cmw}}.
Both JSON and CBOR serializations are allowed in CMW, with the emitter choosing which serialization to use.

> The attestation payload must contain assertions relating to the client's TLS

Reviewer 2: I see it as a strong requirement for the attestation result producer.
Therefore, I’d prefer an upper-case must.

> Identity Key (TIK-C), which associate the private key with the attestation

Author: No, it's always LTK.

> Identity Key (TIK-C), which associate the private key with the attestation

Reviewer 3: > Yaron Sheffer
>  [Aug 30th, 2024 at 11:59 PM](https://ietf.slack.com/archives/C04G8MY5Q1W/p1725055188681899)
> I'm not sure why the draft even needs to mention the DH key. On the other hand, when we discussed TIK in the past it was our consensus that it should be ephemeral (i.e. generated once per session) so I'm not very happy about using the term "ephemeral key" for something else.

In order to remove such and other confusions, we had later agreed to consider EK and LTK rather than TIK.

the X.509 certificate's end entity key.

In the attestation-only mode, the TIK is generated by the TEE and is only used for TLS handshakes.
We RECOMMEND that the TIK be persistent within the TEE. When this is the case,
the TIK may be used by (specialized) applications to provide continuity of identity between TLS sessions with the same peer. 

# Attestation Extensions

As typical with new features in TLS, the client indicates support for the new
extension in the ClientHello message. The newly introduced extensions allow
remote attestation credentials and nonces to be exchanged. The nonces are used
for guaranteeing freshness of the exchanged evidence when the background check
model is in use. Nonces are not used in the passport model, because the expectation
of freshness is more relaxed and is only governed by the lifetime of the signed
attestation results.

When either the Evidence or the Attestation Results extension is successfully
negotiated, the content of the corresponding Certificate message contains a
payload that is encoded based on the wrapper defined in {{-cmw}}.
Both JSON and CBOR serializations are allowed in CMW, with the emitter choosing which serialization to use.

> The attestation payload MUST contain assertions relating to the client's TLS

Reviewer: I'm guessing the answer is "not defined here", but how is this `MUST` enforceable in the case of x509 + attestation? That was generally the purpose of the (now defunct) binder

Author: But even before the binder we had the KAT/PAT construct, which provides the same guarantees.

Yes the KAT/PAT is defined in an unadopted draft and it's presented as "one option". When the dust settles a bit, we may have to switch to a more popular alternative.

## Handshake Overview {#handshake-overview}

The handshake defined here is analogous to certificate-based authentication in a regular TLS handshake.

> We use the TLS Identity Key (TIK) which is either a stand-alone key or is identical

Author: If the new text lacks clarity, I would appreciate text proposals that do not depend on terminology (EK, LTK) defined elsewhere. We should base our terminology on RFC 9334 and the RATS ecosystem.

to the certificate's private key (see {{tik}}).

> This key is attested, with attestation being carried

Author: I'm not a big fan of the term "key attestation" myself. However draft-bft-rats-kat defines it already and we build on that draft: 

> Key Attestation: Evidence containing properties of the environment(s) in which the private keys are stored. For example, a Relying Party may want to know whether a private key is stored in a hardware security module and cannot be exported in unencrypted fashion.

> This key is attested, with attestation being carried

Reviewer: For completeness: key attestation is just a fancy way of (ab)using the way attestation works by extending the chain from workload attestation to a key. Not that different from the way workload attestation builds onto platform attestation. It's explicitly needed here because otherwise there's no way to cryptographically link the credential of a workload to the workload using it, so that you can actually trust the signature found in CertificateVerify based on something.

in the Certificate message. Following that, the peer being attested proves possession of the private key using the CertificateVerify message.

Depending on the use case, the protocol supports peer authentication
using attestation only, or using both attestation and a regular public
key certificate.

The current version of the document assumes the KAT/PAT construct of 
{{I-D.bft-rats-kat}}. Not all platforms support this model, and a document

> that defines private key attestation for use in TLS Attestation as defined here, must specify

Reviewer 2: Ditto, as my previous comment (i.e., s/must/MUST/).

Author: I try to avoid normative language when it applies to future documents rather than implementations. Maybe a matter of taste.

~~~~
{: #figure-extension-results title="TLS Extension Structure for Attestation Results."}

> In the Passport model, both attestation results and an indication of key attestation are

Reviewer 2: The indication of key attestation is part of the attestation results.

Author: I'll keep the text, because the next sentence says we have not defined it.

## Passport Model

The `results_proposal` and `results_request` extensions are used to negotiate

> the attested TLS protocol, and specifically the verifier identities supported by each peer. These

Reviewer 2: Unsure what you mean by “attested TLS protocol”.
Is it a typo for “remote attestation protocol”?

Reviewer: ```suggestion
the supported Attestation Results properties, specifically the verifier identities supported by each peer. These
```
Does this sound better?

Author: @ionut-arm Not really, because the intent was that the TLS extensions actually negotiate the use of this whole protocol. I'll try "the protocol defined here".

TBD.

## Security Guarantees {#sec-guarantees}

> We note that as a pure cryptographic protocol, attested TLS as-is only guarantees that the Identity Key is known by the TEE. A number of additional guarantees must be provided by the platform and/or the TLS stack,

Reviewer 3: Identity Key is LTK or EK?

and the overall security level depends on their existence and quality of assurance:

> * The Identity Key is generated by the TEE.

Reviewer 3: EK, right?

* The Identity Key is never exported or leaked outside the TEE.

> * The TLS protocol, whether implemented by the TEE or a layer above it, is implemented correctly and (for example) does not leak any session key material.

Reviewer 2: Separation doesn’t necessarily mean layering.

I suggest:
```suggestion
* The TLS protocol, whether implemented by the TEE or outside the TEE, is implemented correctly and (for example) does not leak any session key material.
```

> These properties may be explicitly promised ("attested") by the platform, or they can be assured in other ways such as by providing source code, reproducible builds, formal verification etc. The exact mechanisms are out of scope of this document.

Reviewer 2: A couple of missing commas

```suggestion
These properties may be explicitly promised ("attested") by the platform, or they can be assured in other ways, such as by providing source code, reproducible builds, formal verification, etc. The exact mechanisms are out of scope of this document.
```

Author: Some languages are more strict about commas. My personal tendency is similar, but when writing English I adopt the more relaxed American approach.
